src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,111 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final BookRepository bookRepository;
+    private final UserRepository userRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    @Transactional
+    public ShoppingCartResponseDto addToShoppingCart(
+            CartItemRequestDto requestDto,
+            Long userId
+    ) {
+        CartItem cartItem = new CartItem();
+        cartItem.setQuantity(requestDto.getQuantity());
+
+        Long bookId = requestDto.getBookId();
+        Book book = bookRepository.findById(bookId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find book by this ID: " + bookId
+                ));
+        cartItem.setBook(book);
+
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find user by this ID: " + userId
+                ));
+        ShoppingCart shoppingCartFromDB =
+                shoppingCartRepository.findByUserId(userId)
+                        .orElseGet(() -> {
+                            ShoppingCart shoppingCart = new ShoppingCart();
+                            shoppingCart.setUser(user);
+                            shoppingCartRepository.save(shoppingCart);
+                            return shoppingCart;
+                        });
+        cartItem.setShoppingCart(shoppingCartFromDB);
+        cartItemRepository.save(cartItem);
+        shoppingCartFromDB.getCartItems().add(cartItem);
+
+        return shoppingCartMapper.toDto(shoppingCartFromDB);
+    }
+
+    @Override
+    public ShoppingCartResponseDto getShoppingCartByUserId(Long userId) {
+        return shoppingCartMapper.toDto(findShoppingCartByUserId(userId));
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartResponseDto updateCartItem(
+            Long cartItemId,
+            CartItemUpdateRequestDto requestDto,
+            Long userId
+    ) {
+        ShoppingCart shoppingCart = findShoppingCartByUserId(userId);
+        CartItem cartItem =
+                findCartItemByShoppingCartIdAndCartItemId(cartItemId, shoppingCart);
+        cartItem.setQuantity(requestDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public void deleteCartItemById(Long userId, Long cartItemId) {
+        ShoppingCart shoppingCart = findShoppingCartByUserId(userId);
+        CartItem cartItem =
+                findCartItemByShoppingCartIdAndCartItemId(cartItemId, shoppingCart);
+        cartItemRepository.delete(cartItem);
```

Comment:
i assume that you need to create a CartItemService and implement the logic associated with the CartItem

Reply:
Including all the shopping cart logic in one place was my priority, so I did not want to create a separate service for cart items.

Reply:
I’m basing my beliefs on our condition of the task and Sofia’s lectures.